### PR TITLE
[DO NOT MERGE] Composite PR generally about operation cancellation and retrying

### PR DESCRIPTION
**Pull request by mtopolnik**

### CODE REVIEW

**Comment by mtopolnik** [comment]
run-lab-run

**Comment by pveentjer** [comment]
Looks very good!

**Comment by cangencer** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java [hunk 1]
```diff
@@ -178,37 +180,84 @@ public final Operation setReplicaIndex(int replicaIndex) {
...
+     *     <li>When an Invocation of the operation is created, call ID is assigned a positive value.</li>
+     *     <li>When the invocation ends, the operation is {@link #deactivate()}d, but the call ID is preserved.</li>
+     *     <li>The same operation may be involved in a further invocation (retrying); this will assign a
+     *     new call ID and reactivate the operation.</li>
+     * </ul>
+     * @return the call ID
      */
     public final long getCallId() {
-        return callId;
+        return Math.abs(callId);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: Why the abs? call id can't be negative.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: I see it. You are using a negative call id for a special purposes.

Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: Please review https://github.com/hazelcast/hazelcast/pull/9267/commits/a33dd8dc77062ce189583dc7e7283498447cce4d in isolation, this will make things clear.

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java [hunk 2]
```diff
@@ -178,37 +180,84 @@ public final Operation setReplicaIndex(int replicaIndex) {
...
+            throw new IllegalArgumentException("Attempted to set non-positive call ID " + newId);
+        }
+        final long c = this.callId;
+        if (c > 0) {
+            throw new IllegalStateException(String.format(
+                    "Attempt to overwrite the call ID of an active operation: current %d, requested %d",
+                    this.callId, newId));
+        }
+        if (!CALL_ID.compareAndSet(this, c, newId)) {
+            throw new IllegalStateException(String.format("Concurrent modification of call ID. Initially observed %d,"
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: Can you include to toString of the operation in the exception?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java [hunk 3]
```diff
@@ -106,13 +106,13 @@ public final boolean isDone() {
...
 
-    private boolean isDone(final Object state) {
+    private static boolean isDone(final Object state) {
         if (state == null) {
             return true;
         }
 
         return !(state == VOID
-                || state.getClass() == WaitNode.class
+                || state instanceof WaitNode
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java]: why the instanceof check? WaitNode final internal class.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java]: Not terribly important.

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AbstractLockOperation.java [hunk 4]
```diff
@@ -28,19 +28,23 @@
...
 
 import java.io.IOException;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
 
 public abstract class AbstractLockOperation extends Operation
         implements PartitionAwareOperation, IdentifiedDataSerializable {
 
     public static final int ANY_THREAD = 0;
 
+    private static final AtomicLongFieldUpdater<AbstractLockOperation> REFERENCE_CALL_ID =
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/AbstractLockOperation.java]: Ok. Makes sense.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java [hunk 5]
```diff
@@ -317,38 +307,6 @@ public void run0() {
             log(invocationCount, backupTimeouts, normalTimeouts);
         }
 
-        /**
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java]: Nice to see this hack getting resolved

File: hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java [hunk 6]
```diff
@@ -50,6 +50,15 @@ public static void setCallId(Operation op, long callId) {
     }
 
     /**
+     * Resets the callId for the Operation to zero. For details see {@link Operation#resetCallId()}
+     *
+     * @param op the Operation on which to reset the callId
+     */
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java]: Can you add the @returns as well?

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /**
+     * Makes sure that a list of counters is created for a member.
+     * <p/>
+     * This method exists to ensure that an operation-heartbeat is always sent, even if there are no running operations.
+     *
+     * @param address the address of the member.
+     */
+    public void ensureMember(Address address) {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: seems we have two methods doing the same thing?

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    private static long[] toArray(List<Long> longs) {
+        long[] array = new long[longs.size()];
+        for (int k = 0; k < array.length; k++) {
+            array[k] = longs.get(k);
+        }
+        return array;
+    }
+
+    static final class CategorizedCallIds {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: class can be made private

File: hazelcast/src/main/java/com/hazelcast/spi/LiveOperations.java [hunk 9]
```diff
@@ -18,83 +18,12 @@
...
-        }
-
-        return array;
-    }
+public interface LiveOperations {
 
-    public void clear() {
-        callIdsByMember.clear();
-    }
+    void add(Address address, long callId);
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/LiveOperations.java]: some javadoc might be useful here - Saying that this is the caller's address and call id of the operation which is to be tracked.

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java [hunk 10]
```diff
@@ -178,37 +180,85 @@ public final Operation setReplicaIndex(int replicaIndex) {
...
+     */
+    final boolean deactivate() {
+        long c = callId;
+        if (c <= 0) {
+            return false;
+        }
+        if (CALL_ID.compareAndSet(this, c, -c)) {
+            return true;
+        }
+        if (callId > 0) {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: what if callId is set back to 0? or is that impossible? If not, maybe that should fail as well?

**Comment by mtopolnik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java [hunk 11]
```diff
@@ -178,37 +180,85 @@ public final Operation setReplicaIndex(int replicaIndex) {
...
-        return callId;
+        return Math.abs(callId);
+    }
+
+    /**
+     * Tells whether this operation is involved in an ongoing invocation. Such an operation always has a
+     * positive call ID.
+     * @return {@code true} if the operation's invocation is active; {@code false} otherwise
+     */
+    final boolean hasActiveInvocation() {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: might be better to call the method isActive - to match better with "deactivate"

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java [hunk 12]
```diff
@@ -131,30 +139,32 @@ public long getLastCallId() {
...
         @Override
         public int getMaxConcurrentInvocations() {
             return maxConcurrentInvocations;
         }
 
         @Override
         public long next(Invocation invocation) {
-            assert invocation.op.getCallId() == 0 : "callId should be null:" + invocation;
+            final Operation op = invocation.op;
+            assert !invocation.isActive() : "Invocation is already in progress:" + invocation;
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java]: this check seems redundant, since setCallId will already fail later on with an active invocation. The method can be then simplified as `long next(boolean isUrgent)`, furthermore, it's probably better to make it throw a checked exception and handle the exception at the caller site.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java [hunk 13]
```diff
@@ -637,6 +433,180 @@ boolean detectAndHandleBackupTimeout(long timeoutMillis) {
...
+        return allowed;
+    }
+
+    private void invoke0(boolean isAsync) {
+        if (invokeCount > 0) {
+            throw new IllegalStateException("This invocation is already in progress");
+        } else if (isActive()) {
+            throw new IllegalStateException(
+                    "Attempt to reuse the same operation in multiple invocations. Operation is " + op);
+        }
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java]: won't these checks be covered by the setCallId() check later on?

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java [hunk 14]
```diff
@@ -637,6 +433,180 @@ boolean detectAndHandleBackupTimeout(long timeoutMillis) {
...
+
+        invokeCount++;
+
+        // register method assumes this method has run before it is being called so that remote is set correctly
+        if (!initInvocationTarget()) {
+            return;
+        }
+
+        setInvocationTime(op, context.clusterClock.getClusterTime());
+        if (!context.invocationRegistry.register(this)) {
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java]: what happens to invokeCount here? looks like it'll never be decremented?

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public Set<Address> addresses() {
+        return callIdsByMember.keySet();
+    }
+
+    public OperationControl toOpControl(Address address) {
+        CategorizedCallIds callIds = callIdsByMember.get(address);
+        if (callIds == null) {
+            throw new IllegalArgumentException("unknown address");
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: _minor:_ Can you append `address` to the exception message?

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java (hunk 7 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: For clarity. I don't want to call `getOrCreateCallIdsForMember`, ignoring the return value, when all i want is to ensure the address is added to the map.

File: hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package com.hazelcast.spi;
+
+import com.hazelcast.nio.Address;
+
+/**
+ * Implemented by a service that can cancel its operations.
+ */
+public interface CanCancelOperations {
```
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java]: This interface is actually symmetrical to `LiveOperationsTracker` and should be implemented by the same classes that implement that one. However these implementations are out of scope in this PR. The primary motivation is cancellation of Jet jobs and we can gradually add more cancelability to main HZ operations.

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java (hunk 8 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: Done.

File: hazelcast/src/main/java/com/hazelcast/spi/LiveOperations.java (hunk 9 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/LiveOperations.java]: Doen.

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java (hunk 11 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: Originally it was called `isActive` but was renamed to make it more clear. But perhaps I can rename it back because there is an `isActive` method on Invocation now.

File: hazelcast/src/main/java/com/hazelcast/spi/Operation.java (hunk 10 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/Operation.java]: Impossible. Zero is only the starting point and in fact `callId` should never assume the same value twice. The sequence is `0` -> `id1` -> -`id1` -> `id2` -> `-id2` ...

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java (hunk 12 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/CallIdSequence.java]: These are good ideas, I'll implement it that way. About the check, it's  an `assert`, not a business logic-level check. However, after all the changes it does indeed seem redundant.

File: hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java (hunk 16 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java]: I see. 
Currently `InvocationFuture.cancel()` is returning false and doing nothing. With your change, future will be completed with `CancellationException` and by default cancellation (until proper impls are provided) still will be a no-op.
I'm not exactly sure, but won't this be an issue? `future.cancel()` returns `true` but actually does nothing.

Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java]: Actually the current state is broken with respect to `Future` contract: 
> After this method returns, subsequent calls to `isDone()` will always return `true`.
The new state is correct because `Future` is not job-control API; by `cancel`ing it you simply waive your interest in the eventual return value. Taking no action to percipitate task completion is by the book.

File: hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java (hunk 15 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CallsPerMember.java]: Done

File: hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java (hunk 16 above)
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java]: also contract says:
>  If successful, and this task has not started when cancel is called, this task should never run.

Of course this depends on definition of _this task has not started_ ...

Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/CanCancelOperations.java]: Yes, since the task is starting concurrently, this can't be anything more than a hint to the implementation.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java (hunk 14 above)
Comment (mtopolnik) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java]: `register` can also throw a `HazelcastOverloadException`. That case is also unhandled.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java (hunk 13 above)
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java]: OK, I'm considering this out of scope for this PR.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java (hunk 14 above)
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java]: same as previous comment - out of scope of this PR..
